File: java/src/main/java/io/promptresponse/api/TemplateBuilder.java
package io.promptresponse.api;

import io.promptresponse.models.*;
import java.time.Instant;
import java.util.ArrayList;

public class TemplateBuilder {
    private final Metadata metadata;
    private final ArrayList<Section> sections = new ArrayList<>();
    private int sectionCounter = 0;
    private int promptCounter = 0;

    public TemplateBuilder(String title, String templateId) {
        this.metadata = new Metadata();
        this.metadata.setTitle(title);
        this.metadata.setTemplateId(templateId);
        this.metadata.setCreated(Instant.now());
        this.metadata.setModified(Instant.now());
    }

    public TemplateBuilder setDescription(String description) {
        metadata.setDescription(description);
        return this;
    }

    public TemplateBuilder setAuthor(String author) {
        metadata.setAuthor(author);
        return this;
    }

    public TemplateBuilder setVersion(String version) {
        metadata.setTemplateVersion(version);
        return this;
    }

    public SectionBuilder addSection(String title) {
        return addSection(title, null);
    }

    public SectionBuilder addSection(String title, String description) {
        sectionCounter++;
        String sectionId = String.format("section_%03d", sectionCounter);
        Section section = new Section(sectionId, title);
        section.setDescription(description);
        return new SectionBuilder(this, section);
    }

    public AprDocument build() {
        metadata.setModified(Instant.now());
        AprDocument document = new AprDocument();
        document.setVersion("1.0");
        document.setDocumentType(DocumentType.TEMPLATE);
        document.setSections(sections);
        document.setMetadata(metadata);
        return document;
    }

    void addSection(Section section) {
        sections.add(section);
    }

    String nextPromptId() {
        promptCounter++;
        return String.format("prompt_%03d", promptCounter);
    }

    public static class SectionBuilder {
        private final TemplateBuilder templateBuilder;
        private final Section section;

        SectionBuilder(TemplateBuilder templateBuilder, Section section) {
            this.templateBuilder = templateBuilder;
            this.section = section;
        }

        public SectionBuilder addPrompt(String label) {
            return addPrompt(label, null, null, null);
        }

        public SectionBuilder addPrompt(String label, String expectedType, String placeholder, String helpText) {
            String promptId = templateBuilder.nextPromptId();
            Prompt prompt = new Prompt(promptId, label);

            if (expectedType != null || placeholder != null || helpText != null) {
                PromptHints hints = new PromptHints();
                hints.setExpectedDataType(expectedType);
                hints.setPlaceholder(placeholder);
                hints.setHelpText(helpText);
                prompt.setHints(hints);
            }

            section.getPrompts().add(prompt);
            return this;
        }

        public TemplateBuilder done() {
            templateBuilder.addSection(section);
            return templateBuilder;
        }
    }
}

File: python/promptresponse/api/template_builder.py
"""
High-level API for building APR templates programmatically.
"""
from datetime import datetime
from typing import Optional, List
from ..models import (
    AprDocument,
    DocumentType,
    Section,
    Prompt,
    PromptHints,
    Metadata
)


class TemplateBuilder:
    """Builder for creating APR templates programmatically."""

    def __init__(self, title: str, template_id: Optional[str] = None):
        """
        Initialize a new template builder.

        Args:
            title: Template title
            template_id: Optional template identifier
        """
        self.metadata = Metadata(
            title=title,
            template_id=template_id,
            created=datetime.utcnow(),
            modified=datetime.utcnow()
        )
        self.sections: List[Section] = []
        self._section_counter = 0
        self._prompt_counter = 0

    def set_description(self, description: str) -> 'TemplateBuilder':
        """Set template description."""
        self.metadata.description = description
        return self

    def set_author(self, author: str) -> 'TemplateBuilder':
        """Set template author."""
        self.metadata.author = author
        return self

    def set_version(self, version: str) -> 'TemplateBuilder':
        """Set template version."""
        self.metadata.template_version = version
        return self

    def add_section(self, title: str, description: Optional[str] = None) -> 'SectionBuilder':
        """
        Add a new section to the template.

        Args:
            title: Section title
            description: Optional section description

        Returns:
            SectionBuilder for building the section
        """
        self._section_counter += 1
        section_id = f"section_{self._section_counter:03d}"

        section = Section(
            id=section_id,
            title=title,
            description=description,
            prompts=[],
            sections=[]
        )

        section_builder = SectionBuilder(self, section)
        return section_builder

    def build(self) -> AprDocument:
        """
        Build the final APR template document.

        Returns:
            Complete APR template document
        """
        self.metadata.modified = datetime.utcnow()

        return AprDocument(
            version="1.0",
            document_type=DocumentType.TEMPLATE,
            sections=self.sections,
            metadata=self.metadata
        )


class SectionBuilder:
    """Builder for creating sections in a template."""

    def __init__(self, template_builder: TemplateBuilder, section: Section, parent_builder: Optional['SectionBuilder'] = None):
        """
        Initialize section builder.

        Args:
            template_builder: Parent template builder
            section: Section being built
            parent_builder: Parent section builder (for nested sections)
        """
        self.template_builder = template_builder
        self.section = section
        self._parent_builder = parent_builder

    def add_prompt(
        self,
        label: str,
        expected_type: Optional[str] = None,
        placeholder: Optional[str] = None,
        help_text: Optional[str] = None
    ) -> 'SectionBuilder':
        """
        Add a prompt to this section.

        Args:
            label: Prompt label/question
            expected_type: Expected data type (text, email, phone, etc.)
            placeholder: Placeholder text
            help_text: Help text for the prompt

        Returns:
            Self for chaining
        """
        self.template_builder._prompt_counter += 1
        prompt_id = f"prompt_{self.template_builder._prompt_counter:03d}"

        hints = None
        if expected_type or placeholder or help_text:
            hints = PromptHints(
                expected_data_type=expected_type,
                placeholder=placeholder,
                help_text=help_text
            )

        prompt = Prompt(
            id=prompt_id,
            label=label,
            hints=hints
        )

        self.section.prompts.append(prompt)
        return self

    def add_section(self, title: str, description: Optional[str] = None) -> 'SectionBuilder':
        """
        Add a nested section to this section.

        Args:
            title: Nested section title
            description: Optional nested section description

        Returns:
            SectionBuilder for building the nested section
        """
        self.template_builder._section_counter += 1
        section_id = f"section_{self.template_builder._section_counter:03d}"

        nested_section = Section(
            id=section_id,
            title=title,
            description=description,
            prompts=[],
            sections=[]
        )

        return SectionBuilder(self.template_builder, nested_section, parent_builder=self)

    def done(self) -> TemplateBuilder:
        """
        Finish building this section and return to template builder.

        Returns:
            Parent template builder or parent section builder
        """
        if self._parent_builder is not None:
            self._parent_builder.section.sections.append(self.section)
            return self._parent_builder
        self.template_builder.sections.append(self.section)
        return self.template_builder
